src/saveload/details.rs

```diff
@@ -0,0 +1,199 @@
+
+
+
+use std::error::Error;
+use std::fmt::{self, Formatter};
+
+use serde::de::DeserializeOwned;
+use serde::ser::Serialize;
+
+use {Component, Entity, ReadStorage, SystemData, WriteStorage};
+
+use super::marker::Marker;
+
+#[derive(Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct EntityData<M: Marker, E: Error, T: Components<M::Identifier, E>> {
+    pub marker: M,
+    pub components: T::Data,
+}
+
+/// Trivial component is anything that is Component + Copy + DeserializeOwned + Serialize
+pub trait SerializableComponent: Component + DeserializeOwned + Serialize {}
+impl<C> SerializableComponent for C
+where
+    C: Component + DeserializeOwned + Serialize,
+{
+}
+
+/// This trait should be implemented in order to allow component
+/// to be serializeble with `SerializeSystem`.
+/// It is automatically implemented for all `SerializableComponent`s
+pub trait SaveLoadComponent<M>: Component {
+    /// Serializable data representation for component
+    type Data: Serialize + DeserializeOwned;
+
+    /// Error may occur duing serialization or deserialization of component
+    type Error: Error;
+
+    /// Convert this component into serializable form (`Data`) using
+    /// entity to marker mapping function
+    fn save<F>(&self, ids: F) -> Result<Self::Data, Self::Error>
+    where
+        F: FnMut(Entity) -> Option<M>;
+
+    /// Convert this component into deserializable form (`Data`) using
+    /// marker to entity mapping function
+    fn load<F>(data: Self::Data, ids: F) -> Result<Self, Self::Error>
+    where
+        F: FnMut(M) -> Option<Entity>;
+}
+
+/// An error type which cannot be instantiated.
+/// Used as a placeholder for associated error types if
+/// something cannot fail.
+#[derive(Debug)]
+pub enum NoError {}
+
+impl fmt::Display for NoError {
+    fn fmt(&self, _: &mut Formatter) -> fmt::Result {
+        match *self {}
+    }
+}
+
+impl Error for NoError {
+    fn description(&self) -> &str {
+        match *self {}
+    }
+}
+
+impl<C, M> SaveLoadComponent<M> for C
+where
+    C: SerializableComponent + Copy,
+{
+    type Data = Self;
+    type Error = NoError;
+
+    fn save<F>(&self, _ids: F) -> Result<Self::Data, NoError> {
+        Ok(*self)
+    }
+
+    fn load<F>(data: Self, _ids: F) -> Result<Self, NoError> {
+        Ok(data)
+    }
+}
+
+
+
+/// Helper trait defines storages tuples for components tuple
+pub trait Storages<'a> {
+    /// Storages for read
+    type ReadStorages: SystemData<'a> + 'a;
+    /// Storages for write
+    type WriteStorages: SystemData<'a> + 'a;
+}
+
+/// This trait is implemented by any tuple where all elements are
+/// `Component + Serialize + DeserializeOwned`
+pub trait Components<M, E: Error>: for<'a> Storages<'a> {
```

Comment:
No need for the `Error` bound here.

Reply:
But it is required everywhere is is used. Well. Maybe I can just keep it there and remove here.